include/mapbox/variant.hpp

```diff
@@ -645,12 +645,14 @@ class variant
         // move_assign uses move-construction via placement new.
         noexcept(detail::conjunction<std::is_nothrow_move_constructible<Types>...>::value)
     {
+        if (this == &other) return *this; // check for self-assignment
```

Comment:
`other` is an rvalue here.. `(this == &other)` means use-after-move,  do we tolerate that? shall an `assert(false)` be added?

Reply:
Can you elaborate re: use-after-move, please? Are you suggesting we should assert on self-move ?

Reply:
My point is that if `this` got into move assignment it is either temporary or the client code treats it as temporary, so it is not supposed to be used any more.. We could do like
```
if (this == &other) { // playing safe in release mode, hit assertion in debug.
assert(false);
return *this;
}
```

Reply:
Thanks for clarification ^. I'm happy to add `assert` if it helps to avoid self-moving in client code. I'm not totally with you on `this` pointer is being temporary, though. Are you reading this from c++ standard or ..? Sorry, for being dense here :)

Reply:
ok, I see https://stackoverflow.com/questions/9322174/move-assignment-operator-and-if-this-rhs

Reply:
How about just adding assert and removing `if (this == &other)` check as per ^ Howard Hinnant suggestion? 

```c++

assert(this != &other);

```

Reply:
Sorry for not being clear, what I mean is that `other` is either temporary or explicitly moved (being an rvalue), so if `(this == &other)` |=> `this` points to a temporary or a moved value. Of cause, `this` is not itself a temporary in C++ meaning of this term :smile:

Reply:
ok, cool ^ what do you reckon about just having  `assert` ^^

Reply:
https://github.com/mapbox/variant/pull/172/commits/b36c78e120919145adeaf39af0eebd36a1d0feab <-- looks concise and with clear intent to me.

Reply:
I'm a bit concern that self move assignment might still happen in the production code and won't be caught by the assert, so I'd prefer https://github.com/mapbox/variant/pull/172#discussion_r328043024 WDYT?

Reply:
ok done in https://github.com/mapbox/variant/pull/172/commits/3dcac646f10c09a92d5a2808c02e4b0ef29cf413